src/state/include/com/amazonaws/kinesis/video/state/Include.h

```diff
@@ -64,6 +64,17 @@ typedef STATUS (*GetNextStateFunc)(UINT64, PUINT64);
  */
 typedef STATUS (*ExecuteStateFunc)(UINT64, UINT64);
 
+/**
+ * Function which gets called for each state before state machine
+ * transitions to a different state. State machine transitioning to
+ * the same state will not result in the transition hook being called
+ *
+ * @param 1 - IN - Custom data passed in
+ *
+ * @return Status of the callback
+ */
+typedef STATUS (*StateTransitionHookFunc)(UINT64);

```

Comment:
An alternative would be for the ExecuteStateFunc to take a UINT64 which would be the state for which it's called allowing a wrapper API specified in the state machine definition to handle certain logic like retry and then switch/case and call into the state specific execute function.

Reply:
Using a transition hook seems to me a cleaner approach than what you are suggesting